## 오늘의 목표

로마 시대 요리책에서 레시피 카드를 뽑아낸다.


## 오늘의 자료

Apicius, *De re coquinaria* (https://www.gutenberg.org/ebooks/29728)

구텐베르크 프로젝트에서 본문 텍스트 파일을 내려받을 수 있다.

In [1]:
!wget https://www.gutenberg.org/files/29728/29728-0.txt -O apicius.txt

--2022-09-20 08:05:06--  https://www.gutenberg.org/files/29728/29728-0.txt
Resolving www.gutenberg.org (www.gutenberg.org)... 152.19.134.47, 2610:28:3090:3000:0:bad:cafe:47
Connecting to www.gutenberg.org (www.gutenberg.org)|152.19.134.47|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 718740 (702K) [text/plain]
Saving to: ‘apicius.txt’

apicius.txt         100%[===================>] 701.89K  1.36MB/s    in 0.5s    

2022-09-20 08:05:07 (1.36 MB/s) - ‘apicius.txt’ saved [718740/718740]



# 리눅스 기초 명령어

## `wc` (word count)

행의 개수, 단어의 개수, 글자의 개수, 파일명

In [2]:
!wc apicius.txt

 19377 110361 718740 apicius.txt


ASCII 문자의 경우 글자 개수는 파일의 크기(바이트)와 같다. 여기는 718740이다.

In [3]:
!ls -l apicius.txt

-rw-r--r-- 1 root root 718740 Aug 19  2009 apicius.txt


## `head`

기본적으로 파일 내용의 첫 10행을 보여준다.

In [4]:
!head apicius.txt

Project Gutenberg's Cooking and Dining in Imperial Rome, by Apicius

This eBook is for the use of anyone anywhere at no cost and with
almost no restrictions whatsoever.  You may copy it, give it away or
re-use it under the terms of the Project Gutenberg License included
with this eBook or online at www.gutenberg.org


Title: Cooking and Dining in Imperial Rome



## `tail`

기본적으로 파일의 마지막 10행을 보여준다.

In [5]:
!tail apicius.txt



Most people start at our Web site which has the main PG search facility:

     http://www.gutenberg.org

This Web site includes information about Project Gutenberg-tm,
including how to make donations to the Project Gutenberg Literary
Archive Foundation, how to help produce our new eBooks, and how to
subscribe to our email newsletter to hear about new eBooks.


## `cat`

파일 내용 전체를 보여준다. (분량이 많을 때는 웬만하면 하지 말자.)

In [ ]:
!cat apicius.txt

## `|` (pipe)

파일의 51-100행을 보고자 할 때

1. 파일의 첫 100행을 가져온다. (`head -100`)
2. 1의 마지막 50행을 가져온다. (`tail -50`)

In [7]:
!head -100 apicius.txt | tail -50

printed.




                        APICIUS

          COOKERY AND DINING IN IMPERIAL ROME

 A Bibliography, Critical Review and Translation of the
    Ancient Book known as _Apicius de re Coquinaria_

      NOW FOR THE FIRST TIME RENDERED INTO ENGLISH

                          BY
                JOSEPH DOMMERS VEHLING

  _With a Dictionary of Technical Terms, Many Notes,
  Facsimiles of Originals, and Views and Sketches of
     Ancient Culinary Objects Made by the Author_

        INTRODUCTION BY PROF. FREDERICK STARR
       _Formerly of the University of Chicago_




{Transcription:

  APICII LIBRI X

  QVI DICVNTVR DE OBSONIIS
  ET CONDIMENTIS SIUE ARTE
  COQVINARIA QVÆ EXTANT


  NVNC PRIMVM ANGLICE REDDIVIT PROŒMIO
  BIBLIOGRAPHICO ATQVE INTERPRETATIONE
  DEFENSIT UARIISQVE ANNOTATIONIBVS
  INSTRVXIT ITA ET ANTIQVÆ CVLINÆ
  VTENSILIARVM EFFIGIIS EXORNAUIT
  INDICEM DENIQVE ETYMOLOGICVM ET
  TECHNICVM ARTIS MAGIRICÆ ADIECIT


  IOSEPHVS DOMMERS UEHLING

  INTRODVCIT FRIDERICVS ST

In [8]:
!tail -12345 apicius.txt | head -50

    [4] Whether you like pumpkin and brains or not--Apicius
    in this dish reveals himself as the consummate master of
    his art that he really is--a cook for cooks; Moreover,
    the lucidity of his diction in this instance is equally
    remarkable. It stands out in striking contrast to his
    many other formulæ which are so obscured. Many of them
    perhaps were precepts of likewise striking originality
    as this one just cited.


[177] COMPÔTE OF EARLY FRUIT
    _GUSTUM DE PRÆCOQUIS_

CLEAN HARD-SKINNED EARLY FRUITS [1] REMOVE THE SEEDS AND KEEP THEM
COLD IN A PAN. CRUSH PEPPER [2] DRY MINT, MOISTENED WITH BROTH, ADDING
HONEY, RAISIN WINE, WINE AND VINEGAR; POUR THIS OVER THE FRUIT IN THE
PAN, ADDING A LITTLE OIL. STEW SLOWLY ON A WEAK FIRE, THICKEN [the
juice] WITH ROUX [rice flour or other starch diluted with water]
SPRINKLE WITH PEPPER [2] AND SERVE [3].

    [1] Lister praises the early green fruit and the use
    thereof, and, as a physician, recommends imitation of
  

### 해결할 문제

+ 텍스트 파일에는 구텐베르크 프로젝트에 대한 설명 등 본문이 아닌 내용이 포함되어 있다.
+ 서문, 그림 등을 제외하고 레시피에 해당하는 텍스트만 얻고 싶다.
+ 본문에서 살펴본 규칙성은 대략 아래와 같다.

```
[번호] 영어 이름
    _라틴어 이름_

조리법 내용(본문은 대문자, 주석은 소문자로 표기됨)

    [각주번호] 각주
```

+ `cat` 명령에 `-n` 옵션을 넣어서 행 번호와 텍스트를 함께 출력해 보면 몇 행부터 몇 행까지가 레시피에 해당하는지 알 수 있다.

In [ ]:
!cat -n apicius.txt # 2922 # 13673 - 실제로 레시피가 있는 부분

## `>` (redirection)

`[출력될 내용] > [파일명]`

---

아래 코드에서는 `apicius.txt` 파일의 2922행부터 13673행까지를 출력하여 `apicius-text.txt`라는 파일로 쓴다.

In [10]:
!head -13673 apicius.txt | tail +2922 > apicius-text.txt

## `grep`

`grep [옵션] [검색어] [옵션] [파일명]`

---

아래 코드에서는 `apicius-text.txt` 파일에서 `''CABBAGE''`라는 문자열이 매치되는 행을 출력한다.

In [11]:
!grep "CABBAGE" apicius-text.txt

    CHAP.    IX. YOUNG CABBAGE, SPROUTS, CAULIFLOWER.
    CHAP.    XV. SOFT CABBAGE.
[87] YOUNG CABBAGE, SPROUTS [1]
WRAP THE LEEKS WELL IN CABBAGE LEAVES, HAVING FIRST COOKED THEM AS
[103] SOFT CABBAGE
THE CABBAGE IS COOKED WITH POT HERBS IN SODA WATER; PRESS [the water
CELERY, STEWED TENDER GREEN CABBAGE, AND OTHER BOILED GREEN VEGETABLES
HERBS, LEEKS, CORIANDER, DILL, FENNEL, BEETS, MALLOWS, CABBAGE
CABBAGE COOK [separately. Also] CRUSH FENNEL SEED, ORIGANY, SYLPHIUM
PORRIDGE, STIR IT TOGETHER AND USE SOME FINELY CHOPPED CABBAGE STEMS
CORIANDER, DILL, FENNEL, BEETS, MALLOWS, CABBAGE STRUNKS, ALL SOFT AND
GREEN AND VERY FINELY CUT, AND PUT IN A POT. THE CABBAGE COOK
PORRIDGE, STIR, AND USE SOME FINELY CHOPPED CABBAGE STEMS TO SPRINKLE
[469] THE SAME, WITH ANOTHER DRESSING, A CABBAGE _CHARTREUSE_
A QUARTER PART OF LAUREL BERRIES, A MEDIUM-SIZED BOILED CABBAGE,
[Take cooked] MALLOWS, LEEKS, BEETS, OR COOKED CABBAGE SPROUTS [shoots


검색어 앞이나 뒤에 `-n` 옵션을 넣으면 매치된 행 번호가 함께 출력된다.

In [12]:
# -n: line number
!grep -n "CABBAGE" apicius-text.txt

1463:    CHAP.    IX. YOUNG CABBAGE, SPROUTS, CAULIFLOWER.
1469:    CHAP.    XV. SOFT CABBAGE.
1877:[87] YOUNG CABBAGE, SPROUTS [1]
1997:WRAP THE LEEKS WELL IN CABBAGE LEAVES, HAVING FIRST COOKED THEM AS
2156:[103] SOFT CABBAGE
2159:THE CABBAGE IS COOKED WITH POT HERBS IN SODA WATER; PRESS [the water
3019:CELERY, STEWED TENDER GREEN CABBAGE, AND OTHER BOILED GREEN VEGETABLES
3972:HERBS, LEEKS, CORIANDER, DILL, FENNEL, BEETS, MALLOWS, CABBAGE
3974:CABBAGE COOK [separately. Also] CRUSH FENNEL SEED, ORIGANY, SYLPHIUM
3976:PORRIDGE, STIR IT TOGETHER AND USE SOME FINELY CHOPPED CABBAGE STEMS
4802:CORIANDER, DILL, FENNEL, BEETS, MALLOWS, CABBAGE STRUNKS, ALL SOFT AND
4803:GREEN AND VERY FINELY CUT, AND PUT IN A POT. THE CABBAGE COOK
4806:PORRIDGE, STIR, AND USE SOME FINELY CHOPPED CABBAGE STEMS TO SPRINKLE
10098:[469] THE SAME, WITH ANOTHER DRESSING, A CABBAGE _CHARTREUSE_
10102:A QUARTER PART OF LAUREL BERRIES, A MEDIUM-SIZED BOILED CABBAGE,
10126:[Take cooked] MALLOWS, LEEKS, BEETS, OR COO

`-c` 옵션을 넣으면 매치된 행을 출력하지 않고 매치된 횟수만 출력한다.

In [13]:
# -c: count
!grep "PEPPER" -c apicius-text.txt

467


후추가 467번, 꿀이 223번 출현한 것을 확인할 수 있다.

In [14]:
!grep "HONEY" -c apicius-text.txt

223


파이프를 사용하면 첫 번째 문자열의 검색 결과에 한정하여 두 번째 문자열을 검색할 수 있다.

아래 코드에서는 `apicius-text.txt` 파일에서 `"PEPPER"`, `"HONEY"`, `"OIL"`이 모두 포함된 행을 출력한다.

In [15]:
!grep "PEPPER" apicius-text.txt | grep "HONEY" | grep "OIL"

PEPPER, LOVAGE, CORIANDER, RUE, BROTH, HONEY AND A LITTLE OIL.
ANOTHER WAY: THYME, SATURY, PEPPER, LOVAGE, HONEY, BROTH AND OIL.
PUT IN A BRAISIÈRE [2] AND BOIL PEPPER, SPICES, HONEY, BROTH; AND HEAT
WITH OIL, BROTH, REDUCED WINE, WINE, PEPPER, AND HONEY. WHEN DONE
BRAISE THE TRUFFLES WITH PEPPER, MINT, RUE, HONEY, OIL, AND A LITTLE
PEPPER, HERBS, RUE, ONION, HONEY, BROTH, RAISIN WINE, A LITTLE OIL,
PEPPER, CUMIN, RUE, HONEY, VINEGAR, BROTH AND OIL.
CRUSH PEPPER, RUE, A LITTLE HONEY, BROTH, REDUCED WINE, AND OIL TO
LEAVES, PEPPER, HONEY, BROTH, A LITTLE OIL, BIND WITH EGGS IN THE HOT
PEPPER, ORIGANY, ROCKET, MINT, RUE, SAGE [1], DATE WINE, HONEY, OIL,
PEPPER, LOVAGE, RUE, HONEY, NUTS, VINEGAR, WINE, BROTH, A LITTLE OIL;


# awk

`awk [코드] [파일명]`

---

요리책에서 음식 재료를 검색할 때 텍스트의 행 단위보다는 음식 단위로 찾는 것이 자연스럽다. 위에서 사용한 `apicius-text.txt` 파일을 분할하여 한 파일에 하나의 레시피가 들어가게 만들어 보자.

이번에 사용할 `awk` 언어에서는 텍스트를 레코드(행)와 필드(열)로 나누어 준다. 여기에서는 하나의 레시피를 레코드로, 하나의 레시피를 이루는 번호, 영어 이름, 라틴어 이름, 조리법 본문, 각주 등을 각각 필드로 간주할 수 있다. 이를 반영하기 위해 텍스트 내에 레코드와 필드가 각각 어떻게 분할되는지를 컴퓨터에게 알려주어야 한다.

위에서 `apicius-text.txt` 텍스트 파일 본문을 탐색해 보면 각 레시피 사이에는 줄바꿈이 세 번 (혹은 그 이상) 일어나고, 레시피 안에서 이름/본문/각주 사이에는 줄바꿈이 두 번 일어나는 규칙을 발견할 수 있다. 이런 규칙을 반영하여 `RS` 값과 `FS` 값을 지정해 준다.

+ RS(Record separator): 줄바꿈(`\r\n` 또는 `\n`) 세 개 이상의 연쇄
+ FS(Field separator): 줄바꿈 두 개의 연쇄

레코드를 읽어서 화면에 출력할 때는 기존 입력 파일과 다른 구분자를 지정할 수 있다. `ORS`라는 변수명에 구분자로 사용할 문자열을 명세하면 된다.

+ ORS(Output record seperator)

`awk` 언어에서는 코드 블럭을 중괄호 `{ }`로 감싸주면 된다. 코드 블럭 내에서는 세미콜론 `;`으로 명령을 구분해 준다.

`awk` 코드는 `BEGIN{설정}{실행할 내용 1}{실행할 내용 2}{...}`으로 작성한다. (사실 `BEGIN`은 생략해도 된다.)

+ **설정**: 아래의 코드에서는 줄바꿈 문자 세 개 이상의 연쇄를 레코드 구분자로 설정하고, 레코드 내에서는 줄바꿈 문자 두 개의 연쇄를 필드 구분자로 설정한다. 화면에 출력할 때는 하이픈 `-` 다섯 개로 레코드를 구별해 주기로 한다. 
+ **실행할 내용**: `print` 명령으로 각 레코드의 `$1`의 내용을 화면에 출력한다.

`$0`은 하나의 레코드 전체, `$1`은 현재 읽고 있는 레코드 내의 첫 번째 필드를 가리킨다. `apicius-text.txt` 파일에서 `$1`은 `[레시피 번호] 영어 이름\r\n  _라틴어 이름_`에 해당한다.

아래 코드를 실행하면 `-----` 사이에 각 레코드의 첫 번째 필드의 값이 출력된다.

In [16]:
!awk 'BEGIN{RS="\r?\n\r?\n(\r?\n)+"; FS="\r?\n\r?\n"; ORS="\n-----\n"} {print $1}' apicius-text.txt | head -50

I
-----
[1] FINE SPICED WINE
  _CONDITUM PARADOXUM_
-----
[2] HONEY REFRESHER FOR TRAVELERS
  _CONDITUM MELIZOMUM _[1]_ VIATORIUM_
-----
II
-----
[3] ROMAN VERMOUTH
  _ABSINTHIUM ROMANUM_ [1]
-----
III
-----
[4] ROSE WINE [1]
  _ROSATUM_
-----
[5] VIOLET WINE
  _VIOLATIUM_
-----
[6] ROSE WINE WITHOUT [1] ROSES
  _ROSATUM SINE ROSA_
-----
IV
-----
[7] LIBURNIAN OIL
  _OLEUM LIBURNICUM_
-----
V
-----
[8] TO CLARIFY MUDDY WINE
  _VINUM EX ATRO CANDIDUM FACIES_
-----
VI
-----
[9] TO IMPROVE A BROTH [1]
  _DE LIQUAMINE EMENDANDO_ [2]
-----
VII
-----
[10] TO KEEP MEATS FRESH WITHOUT SALT FOR ANY LENGTH OF TIME
   _UT CARNES SINE SALE QUOVIS TEMPORE RECENTES SINT_
-----
[11] TO KEEP COOKED SIDES OF PORK OR BEEF OR TENDERLOINS
   _CALLUM PORCINUM VEL BUBULUM ET UNGUELLÆ COCTÆ UT DIU DURENT_
-----
VIII
-----
[12] TO MAKE SALT MEAT SWEET


위의 내용을 보면 잘된 것과 잘못된 것이 섞여 있다.

+ 잘된 것: `[번호] 영어 이름\n  _라틴어 이름_`
+ 잘못된 것: 페이지 번호 등

잘된 것만 남기기 위해 레시피 형식에 대한 추가 정보를 컴퓨터에게 주고자 한다. 이 데이터에서는 레시피가 `[번호]`로 시작하므로, 첫 번째 필드(`$1`)가 대괄호로 둘러싸인 숫자로 시작한다는 조건을 걸고 싶다. `awk`에서는 실행할 명령 앞에 `if(조건)`을 써 주면 된다.

조건의 내용은 첫 번째 필드(`$1`)의 값이 `[번호]`로 시작한다는 것이다. 정규표현식을 사용해서 이 조건을 명시해 줄 수 있다. `$1~/정규표현식/`으로 쓰면 된다.

아래 코드에서는 두 번째 블럭에 조건문을 적용하여 해당 조건에 맞는 레코드에 대해서만 첫 번째 필드를 출력하도록 한 것이다. 앞의 50행을 살펴보면 조건문을 적용하기 전과 달리 페이지 번호 등이 제외되고 음식 번호와 음식 이름만 잘 나온 것 같다.

In [17]:
!awk 'BEGIN{RS="\r?\n\r?\n(\r?\n)+"; FS="\r?\n\r?\n"; ORS="\n-----\n"} {if($1~/^\[[0-9]+\]/) print $1}' apicius-text.txt | head -50

[1] FINE SPICED WINE
  _CONDITUM PARADOXUM_
-----
[2] HONEY REFRESHER FOR TRAVELERS
  _CONDITUM MELIZOMUM _[1]_ VIATORIUM_
-----
[3] ROMAN VERMOUTH
  _ABSINTHIUM ROMANUM_ [1]
-----
[4] ROSE WINE [1]
  _ROSATUM_
-----
[5] VIOLET WINE
  _VIOLATIUM_
-----
[6] ROSE WINE WITHOUT [1] ROSES
  _ROSATUM SINE ROSA_
-----
[7] LIBURNIAN OIL
  _OLEUM LIBURNICUM_
-----
[8] TO CLARIFY MUDDY WINE
  _VINUM EX ATRO CANDIDUM FACIES_
-----
[9] TO IMPROVE A BROTH [1]
  _DE LIQUAMINE EMENDANDO_ [2]
-----
[10] TO KEEP MEATS FRESH WITHOUT SALT FOR ANY LENGTH OF TIME
   _UT CARNES SINE SALE QUOVIS TEMPORE RECENTES SINT_
-----
[11] TO KEEP COOKED SIDES OF PORK OR BEEF OR TENDERLOINS
   _CALLUM PORCINUM VEL BUBULUM ET UNGUELLÆ COCTÆ UT DIU DURENT_
-----
[12] TO MAKE SALT MEAT SWEET
   _UT CARNEM SALSAM DULCEM FACIAS_
-----
[13] TO KEEP FRIED FISH
   _UT PISCES FRICTI DIU DURENT_
-----
[14] TO KEEP OYSTERS
   _OSTREA UT DIU DURENT_
-----
[15] MAKING A LITTLE LASER GO A LONG WAY
   _UT NUCIA _[1]_ LASERIS TOTO TEM

`awk` 코드를 매번 문자열로 작성하지 않고 파일로 대체할 수 있다. 아래 코드에서는 `echo` 다음에 나오는 문자열을 `apicius.awk`라는 파일로 쓴다. 모든 레코드에 대하여 첫 번째 필드가 `[번호]`로 시작하는 조건을 만족할 때 해당 레코드 전체(`$0`)를 출력하라는 의미다.

In [18]:
!echo 'BEGIN{RS="\r?\n\r?\n(\r?\n)+"; FS="\r?\n\r?\n"; ORS="\n\n\n"} {if($1~/^\[[0-9]+\]/) print $0}' > apicius.awk

파일이 잘 써졌는지를 `cat` 명령으 확인해 보자.

In [19]:
!cat apicius.awk

BEGIN{RS="\r?\n\r?\n(\r?\n)+"; FS="\r?\n\r?\n"; ORS="\n\n\n"} {if($1~/^\[[0-9]+\]/) print $0}


`awk` 명령에 `-f` 옵션을 달아주면 문자열 대신 파일이 코드로 사용된다.

아래의 코드에서는 
1. `apicius.awk`로 작성된 `awk` 코드를
2. `apicius-txt.txt` 텍스트 파일에 적용하여
3. 화면에 출력될 결과를 `apicius-recipes.txt` 파일로 쓴다.

In [20]:
!awk -f apicius.awk apicius-text.txt > apicius-recipes.txt

원래의 `apicius-text.txt` 파일에는 페이지 번호 `I` 등이 들어가 있지만…

In [21]:
!head apicius-text.txt

I


[1] FINE SPICED WINE
  _CONDITUM PARADOXUM_

THE COMPOSITION OF [this] EXCELLENT SPICED WINE [is as follows]. INTO
A COPPER BOWL PUT 6 SEXTARII [1] OF HONEY AND 2 SEXTARII OF WINE; HEAT
ON A SLOW FIRE, CONSTANTLY STIRRING THE MIXTURE WITH A WHIP. AT THE
BOILING POINT ADD A DASH OF COLD WINE, RETIRE FROM STOVE AND SKIM.


`awk` 코드를 적용한 `apicius-recipes.txt`에서는 페이지 번호가 사라졌다.

In [22]:
!head apicius-recipes.txt

[1] FINE SPICED WINE
  _CONDITUM PARADOXUM_

THE COMPOSITION OF [this] EXCELLENT SPICED WINE [is as follows]. INTO
A COPPER BOWL PUT 6 SEXTARII [1] OF HONEY AND 2 SEXTARII OF WINE; HEAT
ON A SLOW FIRE, CONSTANTLY STIRRING THE MIXTURE WITH A WHIP. AT THE
BOILING POINT ADD A DASH OF COLD WINE, RETIRE FROM STOVE AND SKIM.
REPEAT THIS TWICE OR THREE TIMES, LET IT REST TILL THE NEXT DAY, AND
SKIM AGAIN. THEN ADD 4 OZS. OF CRUSHED PEPPER [2], 3 SCRUPLES OF
MASTICH, A DRACHM EACH OF [nard or laurel] LEAVES AND SAFFRON, 5


# 레시피 카드 만들기

지금까지 `awk` 언어를 사용하여 텍스트 파일의 내용을 여러 개의 레코드로 분할해서 처리할 수 있다는 것을 알아보았다. 이제 레시피 하나가 파일 하나에 들어가게 만들어 보자.


`mkdir` 명령으로 `apicius`라는 이름의 폴더를 만든다.

In [23]:
!mkdir apicius

`awk` 코드가 복잡해지므로 한 번에 한 블럭씩 작성해 보자.

`>>`로 꺽쇠를 두 개 사용하면 왼쪽의 출력 결과가 기존 파일의 다음 행에 추가된다.

행별로 하고자 하는 일은 다음과 같다.

1. 레코드 구분자와 필드 구분자를 지정해 준다.
2. 첫 번째 필드(`$1`)의 값은 `[번호] 영어 이름\r\n  _라틴어 이름_`으로 여러 가지 정보를 담고 있다. 이 정보를 분할하기 위해 `split()` 함수를 사용한다. 이 함수는 `split(분할 대상, 분할 결과의 이름, 구분자)`로 작성한다. 해당 블럭을 실행하면 `name`의 첫 번째 요소가 `[번호] 영어 이름`, 두 번째 요소가 `_라틴어 이름_`이 된다.
3. 두 번째 필드(`$2`)의 값은 `레시피 본문(영어 번역)`인데, 종이책의 형식을 그대로 입력했기 때문에 일정한 길이에 따라 줄바꿈이 되어 있다. 이 줄바꿈은 내용을 이해하는 데 불필요하므로 공백 문자로 바꾸어 주자. `gsub(대치 전, 대치 후, 대치 대상)`으로 작성하면 된다.
4. 두 번째 필드의 값, 즉 레시피 본문을 `apicius/[번호] 영어 이름 - _라틴어 이름_.txt`라는 이름의 파일로 쓴다. 레코드 하나마다 파일 하나를 생성하고자 하므로 꺽쇠 `>`를 사용한 재지향을 `awk` 코드 블럭 안에서 사용한다.


In [24]:
# 시작
# 레코드/필드 구분자 지정
!echo 'BEGIN{RS="\r?\n\r?\n(\r?\n)+"; FS="\r?\n(\r?\n)+"}' > apicius-split.awk

# 음식 이름 추출
# 첫 번째 필드(음식 이름)를 줄바꿈 문자와 공백 문자의 연쇄로 분할하여 name이라는 이름으로 저장
!echo '{split($1, name, "(\r?\n)+ *")}' >> apicius-split.awk

# 텍스트 전처리
# 두 번째 필드(조리법 영문 번역)의 줄바꿈 문자를 공백 문자로 대치
!echo '{gsub("(\r?\n)+", " ", $2)}' >> apicius-split.awk

# 파일 쓰기
# 파일명은 name[1](영어 이름)과 name[2](라틴어 이름)을 " - "로 연결
!echo '{if(name[1]!="") print $2 > "apicius/" name[1] " - " name[2] ".txt"}' >> apicius-split.awk

지금 쓴 `awk` 코드가 맞게 작성되었는지 `cat` 명령으로 확인해 보자.

In [25]:
!cat apicius-split.awk

BEGIN{RS="\r?\n\r?\n(\r?\n)+"; FS="\r?\n(\r?\n)+"}
{split($1, name, "(\r?\n)+ *")}
{gsub("(\r?\n)+", " ", $2)}
{if(name[1]!="") print $2 > "apicius/" name[1] " - " name[2] ".txt"}


확인한 후에는 해당 코드를 레시피만 모여 있는 `apicius-recipes.txt` 파일에 대해 실행한다.

In [26]:
!awk -f apicius-split.awk apicius-recipes.txt

`awk` 코드가 잘 실행되었다면 `apicius` 폴더 내에 `ls` 명령을 사용하면 폴더 내의 파일 목록을 볼 수 있다. 간단하게 마지막 10개 파일만 살펴보자.

In [27]:
!ls apicius | tail

[91] ANOTHER WAY - _ALITER_.txt
[92] ANOTHER WAY - _ALITER_.txt
[93] LEEKS - _PORROS_.txt
[94] ANOTHER WAY TO COOK LEEKS - _ALITER PORROS_.txt
[95] ANOTHER WAY - _ALITER PORROS_.txt
[96] LEEKS AND BEANS - _ALITER PORROS_.txt
[97] BEETS - _BETAS_.txt
[98] ANOTHER WAY - _ALITER BETAS ELIXAS_.txt
[99] GREEN VEGETABLES, POT HERBS - _OLISERA_ [1].txt
[9] TO IMPROVE A BROTH [1] - _DE LIQUAMINE EMENDANDO_ [2].txt


텍스트 파일 하나를 골라서 내용을 살펴보면 레시피 하나의 본문이 줄바꿈 없이 잘 저장된 것을 확인할 수 있다.

In [28]:
!cat "apicius/[100] TURNIPS OR NAVEWS - _RAPAS SIVE NAPOS_.txt"

[Turnips are] COOKED [soft, the water is] SQUEEZED [out; then] CRUSH A GOOD AMOUNT OF CUMIN AND A LITTLE RUE, ADD PARTHICAN [1] LASER OR [2] VINEGAR, STOCK, CONDENSED WINE AND OIL [3] HEAT MODERATELY AND SERVE.


# 단어 빈도 세기

`grep` 명령은 하나의 파일이 아닌 폴더에 대해서도 적용할 수 있다. 아래의 코드를 실행하면 `apicius` 폴더 내의 모든 파일 중에서 `"CABBAGE"`을 포함하는 것들이 출력된다. 앞서와 달리 어떤 음식에 양배추가 사용되었는지를 정확하게 알 수 있다.

In [29]:
!grep "CABBAGE" apicius/*

apicius/[103] SOFT CABBAGE - _OLUS MOLLE_.txt:THE CABBAGE IS COOKED WITH POT HERBS IN SODA WATER; PRESS [the water out] CHOP IT VERY FINE: [now] CRUSH PEPPER, LOVAGE, DRY SATURY WITH DRY ONIONS, ADD STOCK, OIL AND WINE.
apicius/[140] A RICH ENTRÉE OF FISH, POULTRY AND SAUSAGE IN CREAM - _PATINA EX LACTE_.txt:SOAK [pignolia] NUTS, DRY THEM, AND ALSO HAVE FRESH SEA-URCHINS [1] READY. TAKE A DEEP DISH [casserole] IN WHICH ARRANGE THE FOLLOWING THINGS [in layers]: MEDIUM-SIZED MALLOWS AND BEETS, MATURE LEEKS, CELERY, STEWED TENDER GREEN CABBAGE, AND OTHER BOILED GREEN VEGETABLES [2], A DISJOINTED [3] CHICKEN STEWED IN ITS OWN GRAVY, COOKED [calf's or pig's] BRAINS, LUCANIAN SAUSAGE, HARD BOILED EGGS CUT INTO HALVES, BIG TARENTINIAN SAUSAGE [4] SLICED AND BROILED IN THE ASHES, CHICKEN GIBLETS OR PIECES OF CHICKEN MEAT. BITS OF FRIED FISH, SEA NETTLES, PIECES OF [stewed] OYSTERS AND FRESH CHEESE ARE ALTERNATELY PUT TOGETHER; SPRINKLE IN BETWEEN THE NUTS AND WHOLE PEPPER, AND THE JUICE AS IS 

아래의 명령은 파일이나 폴더 내의 단어 빈도를 셀 때 전형적으로 사용하는 방법이다. 여기에서는 대문자 로마자의 연쇄(`"[A-Z]+"`)를 단어로 간주했다. 단어의 조건은 데이터에 따라 달라질 수 있다.

+ `grep -h`: 매치된 파일명을 보여주지 않기
+ `grep -o`: 한 행 전체가 아닌 매치되는 문자열만 보여주기
+ `sort`: 알파벳순으로 정렬하기 --- `100`이 `11`보다 먼저 나온다(문자 코드표에서 `0`이 `1`보다 앞에 있으므로).
+ `sort -n`: 수의 크기순으로 정렬하기  --- `11`이 `100`보다 먼저 나온다(수의 크기가 작으므로).
+ `sort -r`: 내림차순으로 정렬하기
+ `uniq`: 같은 요소가 연속으로 나올 때 하나로 합쳐서 보여주기 (`-c`: 연속으로 나온 횟수 보여주기)

In [30]:
!egrep -ho "[A-Z]+" apicius/* | sort | uniq -c | sort -nr > apicius-word-freq.txt

가장 자주 쓰인 단어 30개를 살펴보자. 최상위에는 AND, THE, WITH, A 등 자주 쓰이는 기능어들이 나타나고, 내용어로는 PEPPER, BROTH, WINE, OIL, HONEY, VINEGAR 등이 자주 쓰인다. 고대 로마 사람들의 입맛을 알 수 있다.

In [31]:
!head -30 apicius-word-freq.txt

   1192 AND
    973 THE
    687 WITH
    620 A
    525 IN
    480 OF
    455 PEPPER
    382 BROTH
    338 WINE
    328 OIL
    291 IT
    280 TO
    209 HONEY
    206 ADD
    190 LITTLE
    184 VINEGAR
    184 IS
    179 LOVAGE
    171 SERVE
    167 THIS
    161 WHEN
    154 THEM
    146 SAUCE
    144 CRUSH
    138 OR
    130 PUT
    124 CUMIN
    117 COOKED
    110 SPRINKLE
    101 CORIANDER


위의 상위 30개 목록을 보면 쯔란(CUMIN), 고수(CORIANDER) 등 아시아 음식에서 자주 쓰이는 향신료도 포함되어 있다. 로마 사람들도 양고기를 쯔란으로 양념해 먹었는지 알아보자.

In [32]:
!grep "CUMIN" apicius/* | grep "LAMB" # CUMIN and LAMB

apicius/[355] DAINTY DISHES OF KID OR OF LAMB - _COPADIA HÆDINA SIVE AGNINA_.txt:COOK WITH PEPPER AND BROTH, ALSO WITH VARIOUS ORDINARY BEANS [1] BROTH, PEPPER AND LASER, CUMIN, DUMPLINGS [2] AND A LITTLE OIL [3].
apicius/[356] ANOTHER LAMB STEW - _ALITER HÆDINAM SIVE AGNINAM EXCALDATAM_.txt:PUT [pieces of] KID OR LAMB IN THE STEW POT WITH CHOPPED ONION AND CORIANDER. CRUSH PEPPER, LOVAGE, CUMIN, AND COOK WITH BROTH OIL AND WINE. PUT IN A DISH AND TIE WITH ROUX [1].


쯔란이나 고수가 들어간 음식 목록을 살펴보자. `grep` 대신 `egrep`을 쓰면 정규표현식을 더 쉽게 사용할 수 있다. `-l` 옵션을 쓰면 파일 내용은 나오지 않고 파일 이름만 출력된다.

In [33]:
!egrep -l "CUMIN|CORIANDER" apicius/* # CUMIN or CORIANDER

apicius/[100] TURNIPS OR NAVEWS - _RAPAS SIVE NAPOS_.txt
apicius/[107] FIELD HERBS - _HERBÆ RUSTICÆ_.txt
apicius/[111] A HARMLESS SALAD - _NE LACTUCÆ LÆDANT_.txt
apicius/[113] ANOTHER [Dressing for] CARDOONS - _ALITER CARDUOS_.txt
apicius/[114] BOILED CARDOONS - _ALITER CARDUOS ELIXOS_.txt
apicius/[116] ANOTHER WAY - _ALITER_.txt
apicius/[118] ANOTHER WAY [Purée of Parsnips] [1] - _ALITER_.txt
apicius/[124] ANOTHER WAY - _ALITER_.txt
apicius/[126] APICIAN JELLY - _SALACATTABIA APICIANA_.txt
apicius/[127] OTHER SALACACCABIA - _ALITER_.txt
apicius/[128] EVERYDAY DISH - _PATINA QUOTIDIANA_ [1].txt
apicius/[133] ANOTHER ASPARAGUS CUSTARD - _ALIA PATINA DE ASPARAGIS_.txt
apicius/[137] PUMPKIN PIE - _PATINA DE CUCURBITIS_ [1].txt
apicius/[144] TYROTARICA [1] - _PATELLA THIROTARICA_ [2].txt
apicius/[145] SALT FISH BALLS IN WINE SAUCE [1] - _PATELLA ARIDA_ [2].txt
apicius/[146] VEGETABLE DINNER - _PATELLA EX OLISATRO_ [1].txt
apicius/[154] A FISH STEW - _PATINA ZOMORE_ [1].txt
apicius/[158] SE